🚀 3단계 - 관리자 기능

junseong94 opened the pull request:
안녕하세요 리뷰어님

관리자 기능 구현 완료하여 PR 드립니다.

이 미션을 위해 수정한 내용은 아래와 같습니다.
- Database schema 수정
  - memeber table : role column 추가
- Member entity와 연관 dto 에 role 필드 추가
- Member role 검증을 위한 MemberHandlerInterceptor 구현
  - HandlerInterceptor 를 상속, preHandle() @Override 하여 토큰에서 회원정보를 조회 후 권한 확인하는 로직 구현
  - 이를 위해 AuthService - findfindMember() 와 MemberRepository - findMemberById(long id) 신규 개발
  - addInterceptors() 메서드를 이용하여 `/admin/**` 호출 시 MemberHandlerInterceptor를 호출하도록 등록

이런 경우는 테스트 코드를 어떻게 작성해야할지 모르겠어서 테스트 코드는 작성하지 못했습니다. :(

이상입니다.

junseong94 committed:
rebase(project) : 프로젝트 리베이스

PR된 원격저장소 기준 rebase 수행

junseong94 committed:
rebase : 프로젝트 rebase

PR된 원격저장소 기준 rebase 수행

junseong94 committed:
rebase : 프로젝트 rebase

PR된 원격저장소 기준 rebase 수행

junseong94 committed:
chore(class) : 불필요한 클래스 파일 삭제

- rebase 과정에서 불필요한 남아있는 클래스 파일 삭제

junseong94 committed:
chore : step2 미션 rebase

- step2 미션 기반 Rebase

junseong94 committed:
chore : step2 미션 rebase

- step2 미션 기반 Rebase

junseong94 committed:
feat : 로그인 계정 권한에 따른 url 접근 권한 설정

- member 에 role 추가
  - member 테이블에 column 추가 및 insert query 추가
  - MemberResponseDto, Member 엔티티에 role 필드 추가
- HandlerInterceptor를 상속받아 preHandler를 이용, role 확인 로직 구현
  - AuthService - findMember() 메서드 구현
  - MemberRepository - rowMapper에 role 추가 및 findMemberById(long id) 메서드 구현
- MemberConfiguration 의 addInterceptors() 메서드로 "/admin/**" url 접근 시 MemberHandlerInterceptor를 호출하도록 Interceptor 등록

pkeugine commented on src/main/java/roomescape/member/infra/MemberHandlerInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+    }
+
+    @Override
+    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) throws IOException {
```
```suggestion
    public boolean preHandle(HttpServletRequest request, HttpServletResponse response, Object handler) {
```

사용하지 않는 throws 는 지워주면 좋을 것 같네요!

pkeugine commented on src/main/java/roomescape/member/infra/MemberHandlerInterceptor.java:
```diff
@@ -0,0 +1,41 @@
+
+        LOGGER.info("check role of member : {}", member);
+
+        if (member == null || !member.getRole().equals("ADMIN")) {
```
```suggestion
        if (member == null || !"ADMIN".equals(member.getRole())) {
```

소소한 디테일인데, 이 로직을 위와 같이 작성하면 어떤 장점이 있을까요? :)

junseong94 replied:
member.getRole()은 해당 값이 null 일경우 NPE 가 발생할 수 있는데.
후자의 경우 member.getRole()이 null 이면, false를 반환해 NPE가 발생하지 않네요 :)

이렇게 또 하나 배워가네요 ㅎㅎ

```java
public class test {

    public static void main(String[] args) {
        String answer = "hi";
        System.out.println("hi".equals(answer));

        String answer2 = null;
        System.out.println(answer2.equals("hi"));
    }
}
```

![image](https://github.com/next-step/spring-roomescape-auth/assets/105576721/1a4dcfe1-8b6f-4fb4-b407-fe7bb15f0971)

pkeugine commented on src/main/java/roomescape/member/application/AuthService.java:
```diff
@@ -45,4 +45,14 @@ public MemberResponseDto findMemberName(final String token) {
+        if (!isTokenExpired) {
+            throw new AuthorizationException("만료된 토큰입니다.");
+        }
+        final String id = jwtTokenProvide.extractMemberIdFromToken(token);
```
저도 이전에 못 보고 지나쳤는데, `jwtTokenProvider` 의 r 이 빠졌네요 😂

pkeugine commented on src/main/java/roomescape/member/application/AuthService.java:
```diff
@@ -45,4 +45,14 @@ public MemberResponseDto findMemberName(final String token) {
     }
+
+    public MemberResponseDto findMember(final String token){
+        final boolean isTokenExpired = jwtTokenProvide.validateToken(token);
+        if (!isTokenExpired) {
+            throw new AuthorizationException("만료된 토큰입니다.");
+        }
```
해당 변수의 이름이 상태를 잘못 나타내고 있는 것 같아요.
`isTokenExpired` 라면 '만료된 토큰' 이라는 뜻이라는 생각이 들어서요!
변수명을 반대로 지어보면 좋을 것 같네요!

pkeugine commented on src/main/java/roomescape/member/application/AuthService.java:
```diff
@@ -45,4 +45,14 @@ public MemberResponseDto findMemberName(final String token) {
         final String nameById = memberRepository.findNameById(Long.parseLong(id));
```
커밋 기록을 보니 rebase 관련 어려움을 조금 겪으셨던 것 같아요.
어떤 문제가 있었나요?
이를 파악하면 나중에 똑같은 문제가 생겼을 때 불필요한 커밋 기록을 남기지 않고 해결 가능할 것 같아 질문해봅니다!

junseong94 replied:
로컬 default branch(모든 미션을 rebase하는 branch)를 그동안 rebase를 하지 않고, step1 -> step2 브랜치로 바로 넘어갔었습니다.
이번 rebase 시 default branch는  클래스파일을 도메인 디렉토리안에 구분하지 않고 모아놨었는데, nextstep 원격저장소에는 디렉토리별로 클래스파일을 나눴어서, 과거 클래스파일이 남아있었습니다. 

이후에는 default branch에 지속적으로 rebase를 해서 이러한 이슈가 없도록 하겠습니다.

pkeugine reviewed (changes requested):
안녕하세요 준성님,
이번 스텝은 워낙 깔끔하게 기능구현을 해주셔서 오히려 제가 리뷰를 남길 부분이 많이 없었어요.
그래서 자잘한 리뷰와 rebase 관련 이슈를 겪으신 것 같아 그 관련 질문 남겨봤습니다.
수정할 부분이 많이 없어서 몇 개만 바꿔주시고 다시 리뷰 요청 보내주시면
바로 merge 하고 다음 미션으로 넘어가보죠!
고생하셨습니다 😄 😄 😄

junseong94 committed:
chore : 오타 및 변수명 수정, NPE 예방을 위해 문자열 검증 로직 수정, 사용하지 않는 exception 제거

- `jwtTokenProvide` to `jwtTokenProvider`
- `isTokenExpired` to `isTokenValid`
- NPE 예방을 위해 문자열 검증 로직 수정
  - `!member.getRole().equals("ADMIN")` to `!"ADMIN".equals(member.getRole()`

pkeugine reviewed (approved):
고생하셨습니다 준성님!
바로 반영해주셨네요 :)
머지하고, 다음 미션으로 넘어가보죠!!!
